BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch

```diff
@@ -0,0 +1,213 @@
++            Class<?> ret = handlerMethod.getReturnType();
++            if (ret != void.class) {
++                if (ret == long.class || ret == double.class) {
++                    // https://docs.oracle.com/javase/specs/jvms/se9/html/jvms-2.html#jvms-2.11.1
++                    mv.visitInsn(Opcodes.POP2);
++                } else {
++                    mv.visitInsn(Opcodes.POP);
++                }
++            }
++            mv.visitInsn(Opcodes.RETURN);
++            mv.visitMaxs(0, 0);
++            mv.visitEnd();
++        }
++
++        cw.visitEnd();
++        return cw.toByteArray();
++    }
++
++    private static Class<?> defineClass(ClassLoader loader, String name, byte[] data) {
++        try {
++            return (Class<?>) defineClassMethod.invoke(loader, name, data, 0, data.length);
++        } catch (Exception e) {
++            throw new RuntimeException("Failed to define class '" + name + "'", e);
++        }
++    }
++
++    static {
++        try {
++            defineClassMethod = ClassLoader.class.getDeclaredMethod("defineClass", String.class, byte[].class, int.class, int.class);
++            defineClassMethod.setAccessible(true);
++        } catch (Exception e) {
++            throw new Error("Failed to get ClassLoader#defineClass method", e);
++        }
++    }
++}
+\ No newline at end of file
+diff --git a/event/src/main/java/io/github/waterfallmc/waterfall/impl/event/EventExecutor.java b/event/src/main/java/io/github/waterfallmc/waterfall/impl/event/EventExecutor.java
+new file mode 100644
+index 00000000..64c407c4
+--- /dev/null
++++ b/event/src/main/java/io/github/waterfallmc/waterfall/impl/event/EventExecutor.java
+@@ -0,0 +1,18 @@
++package io.github.waterfallmc.waterfall.impl.event;
++
++import java.lang.reflect.Method;
++import java.lang.reflect.Modifier;
++
++/**
++ * @author Mark Vainomaa
++ */
++public interface EventExecutor {
++    void execute(Object listener, Object event);
++
++    static EventExecutor create(Method method) {
++        if (!Modifier.isPublic(method.getModifiers())) {
++            throw new IllegalArgumentException("Event handler method must be public");
++        }
++        return ASMEventExecutor.getEventExecutor(method);
++    }
++}
+diff --git a/event/src/main/java/net/md_5/bungee/event/EventBus.java b/event/src/main/java/net/md_5/bungee/event/EventBus.java
+index d7c3b90d..2aa4941d 100644
+--- a/event/src/main/java/net/md_5/bungee/event/EventBus.java
++++ b/event/src/main/java/net/md_5/bungee/event/EventBus.java
+@@ -51,10 +51,10 @@ public class EventBus
+                 } catch ( IllegalArgumentException ex )
+                 {
+                     throw new Error( "Method rejected target/argument: " + event, ex );
+-                } catch ( InvocationTargetException ex )
++                } catch ( Exception ex ) // Waterfall - Use ASM to invoke event handlers
```

Comment:
Though this makes me think. What happens when `Method#invoke(...)` throws `Throwable` instead of an `Exception`? Javadoc tells me only about [exception](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/Method.html#invoke-java.lang.Object-java.lang.Object...-).

This needs to be verified, if Throwable was passed on wrapped in InvocationTargetException, then I need to change this catch to handle Throwable instead.